src/ParallelTempering.jl

```diff
@@ -0,0 +1,168 @@
+end
+
+
+"""
+Propose and accept/reject a replica exchange.
+-> Swap spin configurations instead of the β's for now 
+(no bookkeeping or master proc. needed).
+-> Use deterministic even/odd scheme (irreversible).
+"""
+function replica_exchange!(replica::Replica)
+    rex_ID = replica.rex_dir+1
+
+    # first and last replicas only exch. in one dir.
+    if replica.nn_ranks[rex_ID] < 0
+        return false
+    end
+
+    rex_rank = replica.nn_ranks[rex_ID]
+    E_curr = [running_energy(replica.sampler)]
+    rex_accept = [false]
+
+    # propose replica exch.
+    if replica.rank < rex_rank
+        # send energy to neighbor
+        MPI.Send(E_curr, rex_rank, 1, MPI.COMM_WORLD)
+
+        # receive accept/reject info from neighbor
+        MPI.Recv!(rex_accept, rex_rank, 2, MPI.COMM_WORLD)
+    else
+        # receive energy from neighbor
+        E_rex = [0.0]
+        MPI.Recv!(E_rex, rex_rank, 1, MPI.COMM_WORLD)
+
+        # replica exch. acceptance probability
+        P_rex = exp( (replica.nn_βs[rex_ID] - replica.sampler.β)*(E_rex[1] - E_curr[1]) )
+
+        # acceptance criterion.
+        if (P_rex > 1.0) || (rand() <= P_rex)
+            rex_accept[1] = true
+        end
+
+        # send accept/reject info to neighbor
+        MPI.Send(rex_accept, rex_rank, 2, MPI.COMM_WORLD)
+    end
+
+    # replica exch. rejected
+    if !rex_accept[1]
+        return false
+    end
+
+    # accept replica exch.: swap configurations
+    MPI.Sendrecv!(replica.sampler.system.sites, rex_rank, 3,
+                  replica.sampler.system.sites, rex_rank, 3, MPI.COMM_WORLD)
+
+    # recalculate energy of new state
+    replica.sampler.E = energy(replica.sampler.system)
+
+    return true
+end
+
+
+"""
+Start a parallel tempering (PT) simulation. 
+Run w/ the command "mpiexec -n [n_procs] julia --project [julia_script.jl]".
+"""
+function run!(replica::Replica, T_sched::Function;
+                therm_mcs=1000, measure_interval=10, rex_interval=100, max_mcs=500_000, bin_size=1.0)
+    # set replica sampling β using temperature (β = 1.0/(kT)) 
+    T = T_sched(replica.rank+1, replica.N_ranks)
+    set_temp!(replica.sampler, T)
+
+    # set replica neighbor β's  
+    for i in 1:2
+        if replica.nn_ranks[i] != -1
+            replica.nn_βs[i] = 1.0 / T_sched(replica.nn_ranks[i]+1, replica.N_ranks)
+        end
+    end
+
+    # equilibrate replica to it's distribution
+    thermalize!(replica.sampler, therm_mcs)
+
+    N_rex = max_mcs / rex_interval
+    N_measure = cld(rex_interval, measure_interval)
+    replica.sampler.nsweeps = measure_interval
+
+    A = BinnedArray{Float64, Int64}(bin_size=bin_size)
+    rex_accepts = zeros(Int64, 2)
+
+    # start PT with finite length
+    for i in 1:N_rex
+
+        # measurements between exchanges
+        for j in 1:N_measure
+            # decorrelation steps
+            sample!(replica.sampler)
+
+            # add measurements here 
+            #...
```

Comment:
This could be solved in a future pull. Ideally, a typical user wouldn't have to edit any Sunny. One idea to get arbitrary measurements in from the user:

`Replica` could take a `Function` called `measure`, where calling `measure(system)` returns a `Dict` of measurements mapping a string (the measurement name) to whatever form the measurement takes. `Replica` would then call this function right where you currently have `# add measurements here`, and then somehow keep track of all of the measurements internally in a big `Dict`. You could use `Serialization` to `serialize` it out into a big file at the end.

Users can then get whatever measurements they want by passing a custom measurement function.

Not sure what the pitfalls of this approach would be though.

Reply:
Yeah, this would be nice to have some way to store arbitrary observables in each replica, like the dict you mention.